lib/rgb_matrix/animation/config.ex

```diff
@@ -97,42 +156,85 @@ defmodule RGBMatrix.Animation.Config do
       ...>   hue_wave_config,
       ...>   RGBMatrix.Animation.HueWave.Config.schema(),
-      ...>   %{"direction" => "left"}
-      ...> )
+      ...>   %{"direction" => "left"})
       %RGBMatrix.Animation.HueWave.Config{direction: :left, speed: 4, width: 20}
   """
-  @spec update_config(config :: t, schema :: any, params :: map) :: t
+  @spec update_config(
+          config :: t,
+          schema :: config_schema,
+          params :: update_params
+        ) :: t
   def update_config(config, schema, params) do
-    Enum.reduce(params, config, &update_field(&1, &2, schema))
+    params
+    |> Enum.reduce(config, fn param, config ->
+      cast_and_update_field(param, config, schema)
+    end)
   end
 
-  defp create_atom_key(key) when is_binary(key), do: String.to_existing_atom(key)
-  defp create_atom_key(key) when is_atom(key), do: key
+  @spec cast_and_update_field(
+          param :: {atom | String.t(), any},
+          config :: t,
+          schema :: config_schema
+        ) :: t
+  defp cast_and_update_field({key, value} = _param, config, schema) do
+    with key <- create_atom_key(key),
+         {:ok, %type_module{} = type} <- Keyword.fetch(schema, key),
+         {:ok, value} <- type_module.cast(type, value) do
+      Map.put(config, key, value)
+    else
+      error ->
+        field_warn(error, key, value)
+        config
+    end
+  end
 
-  defp create_struct!(map, module), do: struct!(module, map)
+  @spec validate_field(schema_field, creation_params) ::
+          {atom, FieldType.field()} | nil
+  defp validate_field({key, %type_module{} = type} = _field, params) do
+    value = Map.get(params, key, type.default)
 
-  defp update_field({key, value}, config, schema) do
-    key = create_atom_key(key)
+    case type_module.validate(type, value) do
+      :ok ->
+        {key, value}
 
-    %type_module{} = type = Keyword.fetch!(schema, key)
-    {:ok, value} = type_module.cast(type, value)
-    if type_module.validate(type, value) == :error, do: value_error!(value, key)
+      error ->
+        field_warn(error, key, value)
+        {key, type.default}
+    end
+  end
+
+  @spec create_atom_key(key :: atom | String.t()) ::
+          (key :: atom) | :invalid_field
+  defp create_atom_key(key) when is_atom(key), do: key
 
-    Map.put(config, key, value)
+  defp create_atom_key(key) when is_binary(key) do
+    String.to_existing_atom(key)
+  rescue
+    ArgumentError ->
+      :invalid_field
   end
 
-  defp validate_field({key, %type_module{} = type}, params) do
-    value = Map.get(params, key, type.default)
+  @spec create_struct!(map, module) :: struct
+  defp create_struct!(map, module), do: struct!(module, map)
 
-    case type_module.validate(type, value) do
-      :ok -> {key, value}
-      :error -> value_error!(value, key)
-    end
+  @spec field_warn(
+          error :: field_error,
+          key :: atom | String.t(),
+          value :: any
+        ) :: :ok
+  defp field_warn(:cast_error, key, value) do
```

Comment:
I would need to create a wrapper for `Keyword.fetch/2`, then, because that particular function returns `:error` on errors. It should be worth it, though, because then the error could contain more information.

Reply:
Wrapping `Keyword.fetch` to return a different error sounds like a code smell to me. I'm not very familiar with the config code, so what's going through my mind is:
- In what cases would the keyword not exist?
- Are keywords not existing a normal case that we need to handle gracefully?
- Is this a place we should be accessing the keyword list with `schema[key]` so that it returns `nil` instead of `:error`?
  - Should we send a `nil` type to `cast()` and let it produce the correct error if there is no type?
  - Should fetching the keyword (type) be done inside the `with` clause? Should it be pulled into a `case` to do its own error handling?
  - Should the keyword fetch be combined with `create_atom_key()`, turning it into a function like `get_type(key)`?
- If non-existent keywords are abnormal, should we use the bang function `Keyword.fetch!` and let it crash?

You don't have to give me the answers to all of these, but these are the kinds of questions I would be thinking about when refactoring this code.